src/lib/Libutil/pbs_aes_encrypt.c

```diff
@@ -150,3 +152,100 @@ pbs_decrypt_pwd(char *crypted, int credtype, size_t len, char **uncrypted)
 
         return 0;
 }
+
+/**
+ * @brief
+ *	encode_to_base64 - Encode data into base64 format
+ *
+ * @param[in]		buffer			-	Data buffer for encoding
+ * @param[in]		buffer_len		-	Length of the data buffer
+ * @param[in/out]	ret_encoded_data	-	Return the encoded data
+ *
+ * @return int
+ * @retval 0 - for success
+ * @retval 1 - for failure
+ */
+int
+encode_to_base64(const unsigned char* buffer, size_t buffer_len, char** ret_encoded_data)
+{
+	BUF_MEM *b_mem;
+	BIO *mem_obj1, *mem_obj2;
+	int buf_len = 0;
+
+	mem_obj1 = BIO_new(BIO_s_mem());
+	mem_obj2 = BIO_new(BIO_f_base64());
+
+	mem_obj1 = BIO_push(mem_obj2, mem_obj1);
+	BIO_set_flags(mem_obj1, BIO_FLAGS_BASE64_NO_NL);
+	BIO_write(mem_obj1, buffer, buffer_len);
+	(void)BIO_flush(mem_obj1);
+	BIO_get_mem_ptr(mem_obj1, &b_mem);
+
+	buf_len = (*b_mem).length;
```

Comment:
I think right way of access data and length in BIO_MEM is to use `BIO_get_mem_data()`

Reply:
one step above at line 182 we are retrieving a BIO_MEM pointer using BIO_get_mem_ptr() call. is it not a valid call ?

Reply:
Its valid call but I don't think you need that, you can directly do `buf_len = BIO_get_mem_data(mem_obj1, &buf)` define `char *buf`

Reply:
> Its valid call but I don't think you need that, you can directly do `buf_len = BIO_get_mem_data(mem_obj1, &buf)` define `char *buf`

If it is valid, we do not need to change. The only thing we missed is the check for BIO_new() failing, and adding that is sufficient

Reply:
> > Its valid call but I don't think you need that, you can directly do `buf_len = BIO_get_mem_data(mem_obj1, &buf)` define `char *buf`
> 
> If it is valid, we do not need to change. The only thing we missed is the check for BIO_new() failing, and adding that is sufficient

Actually, i take it back. Let's fix this also. I am just worried we have to do another test, but i think this is safe enough

Reply:
@subhasisb Generally `BIO_get_mem_ptr` is used with `BIO_set_mem_buf()`, and accessing .length and .data is not documentented in OpenSSL docs, it says to use `BIO_get_mem_data()` to data from BUF_MEM structure (https://www.openssl.org/docs/man1.0.2/man3/BIO_set_mem_buf.html). But up to you guys...

Reply:
> @subhasisb Generally `BIO_get_mem_ptr` is used with `BIO_set_mem_buf()`, and accessing .length and .data is not documentented in OpenSSL docs, it says to use `BIO_get_mem_data()` to data from BUF_MEM structure (https://www.openssl.org/docs/man1.0.2/man3/BIO_set_mem_buf.html). But up to you guys...

Yes the documentation is insufficient. However, consider this. BIO_get_mem_ptr() is a first class API documented in Openssl docs. It is supposed to return BUF_MEM structure. Therefore the structure is documented and you CAN access it's elements. Consider this snippet of code in the openssl docs https://linux.die.net/man/3/bio_s_mem  
BUF_MEM *bptr;
BIO_get_mem_ptr(mem, &bptr);
BIO_set_close(mem, BIO_NOCLOSE); /* So BIO_free() leaves BUF_MEM alone */
BIO_free(mem);

What do you do with bptr now, if you cannot access its contents?

Reply:
I am not denying that we can't access .length and .data in BUF_MEM, I am just saying we have a better way (using documented API) of doing what we are trying to do here.
But as I said, it's up to you guys 😄

Reply:
> I am not denying that we can't access .length and .data in BUF_MEM, I am just saying we have a better way (using documented API) of doing what we are trying to do here.
> But as I said, it's up to you guys 😄

Ok so that is a valid access, you agree? I thought you are saying the access is invalid :)

Reply:
Yes, accessing .length and .data in BUF_MEM is absolutely valid.